File: ts/src/data/datasets.ts
import type { ReleaseArchiveSpec, ReleaseSpec } from "./sync.js";
import { GAMEDATA_FILES } from "./sync.js";
import AdmZip from "adm-zip";

export const STORYJSON_REQUIRED_FILES = [
  "zh_CN/gamedata/excel/story_review_table.json",
  "zh_CN/storyinfo.json",
] as const;
const STORYJSON_REVIEW_TABLE = "zh_CN/gamedata/excel/story_review_table.json";

export const LEVELS_REQUIRED_FILES = [
  "zh_CN/gamedata/levels/enemydata/enemy_database.json",
] as const;

export interface ReleaseDatasetSpec {
  datasetId: string;
  owner: string;
  repo: string;
  assetName: string;
  requiredFiles: readonly string[];
  validateZip?: (zipPath: string) => string[];
}

export const GAMEDATA_EXCEL: ReleaseDatasetSpec = {
  datasetId: "gamedata.excel",
  owner: "3aKHP",
  repo: "ArknightsGameData",
  assetName: "zh_CN-excel.zip",
  requiredFiles: GAMEDATA_FILES,
};

export const STORY_ZH_CN: ReleaseDatasetSpec = {
  datasetId: "story.zh_CN",
  owner: "3aKHP",
  repo: "ArknightsStoryJson",
  assetName: "zh_CN.zip",
  requiredFiles: STORYJSON_REQUIRED_FILES,
  validateZip: validateStoryjsonZip,
};

export const GAMEDATA_LEVELS: ReleaseDatasetSpec = {
  datasetId: "gamedata.levels",
  owner: "3aKHP",
  repo: "ArknightsGameData",
  assetName: "zh_CN-levels.zip",
  requiredFiles: LEVELS_REQUIRED_FILES,
};

export function releaseSpecForDataset(
  dataset: ReleaseDatasetSpec,
  localZip: string,
): ReleaseSpec {
  return {
    owner: dataset.owner,
    repo: dataset.repo,
    assetName: dataset.assetName,
    localZip,
    validateZip: dataset.validateZip,
  };
}

export function archiveSpecForDataset(
  dataset: ReleaseDatasetSpec,
  localZip: string,
  localRoot: string,
): ReleaseArchiveSpec {
  return {
    owner: dataset.owner,
    repo: dataset.repo,
    assetName: dataset.assetName,
    localZip,
    localRoot,
    requiredFiles: dataset.requiredFiles,
  };
}

function missingEntries(zip: AdmZip, requiredFiles: readonly string[]): string[] {
  const names = new Set(zip.getEntries().map((entry) => entry.entryName));
  return requiredFiles.filter((path) => !names.has(path));
}

function storyPath(storyKey: string): string {
  return `zh_CN/gamedata/story/${storyKey}.json`;
}

export function validateStoryjsonZip(zipPath: string): string[] {
  const zip = new AdmZip(zipPath);
  const missing = missingEntries(zip, STORYJSON_REQUIRED_FILES);
  if (missing.includes(STORYJSON_REVIEW_TABLE)) return missing;

  let table: Record<string, { infoUnlockDatas?: Array<{ storyTxt?: string }> }>;
  try {
    const entry = zip.getEntry(STORYJSON_REVIEW_TABLE);
    table = JSON.parse(entry!.getData().toString("utf-8")) as typeof table;
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return [...missing, `${STORYJSON_REVIEW_TABLE} is unreadable: ${message}`];
  }

  const names = new Set(zip.getEntries().map((entry) => entry.entryName));
  const storyPaths = [...new Set(
    Object.values(table)
      .flatMap((entry) => entry.infoUnlockDatas ?? [])
      .map((data) => data.storyTxt)
      .filter((storyKey): storyKey is string => Boolean(storyKey))
      .map(storyPath),
  )].sort();

  return [...missing, ...storyPaths.filter((path) => !names.has(path))];
}

File: python/src/prts_mcp/data/datasets.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from zipfile import BadZipFile, ZipFile

from prts_mcp.data.sync import GAMEDATA_FILES, ReleaseArchiveSpec, ReleaseSpec


STORYJSON_REQUIRED_FILES: tuple[str, ...] = (
    "zh_CN/gamedata/excel/story_review_table.json",
    "zh_CN/storyinfo.json",
)
STORYJSON_REVIEW_TABLE = "zh_CN/gamedata/excel/story_review_table.json"
LEVELS_REQUIRED_FILES: tuple[str, ...] = (
    "zh_CN/gamedata/levels/enemydata/enemy_database.json",
)


@dataclass(frozen=True)
class ReleaseDatasetSpec:
    dataset_id: str
    owner: str
    repo: str
    asset_name: str
    required_files: tuple[str, ...]

    def release_spec(self, local_zip: Path) -> ReleaseSpec:
        return ReleaseSpec(
            owner=self.owner,
            repo=self.repo,
            asset_name=self.asset_name,
            local_zip=local_zip,
            validate_zip=self.validate_zip,
        )

    def archive_spec(self, *, local_zip: Path, local_root: Path) -> ReleaseArchiveSpec:
        return ReleaseArchiveSpec(
            owner=self.owner,
            repo=self.repo,
            asset_name=self.asset_name,
            local_zip=local_zip,
            local_root=local_root,
            required_files=self.required_files,
        )

    def validate_zip(self, zip_path: Path) -> list[str]:
        """Return missing or invalid zip entries for this dataset."""
        try:
            with ZipFile(zip_path) as zf:
                return validate_storyjson_zip(zf) if self is STORY_ZH_CN else _missing_entries(zf, self.required_files)
        except BadZipFile:
            return [f"{zip_path} is not a valid zip"]


def _missing_entries(zf: ZipFile, required_files: tuple[str, ...]) -> list[str]:
    names = set(zf.namelist())
    return [path for path in required_files if path not in names]


def _story_path(story_key: str) -> str:
    return f"zh_CN/gamedata/story/{story_key}.json"


def validate_storyjson_zip(zf: ZipFile) -> list[str]:
    """Validate story metadata and referenced chapter JSON entries."""
    missing = _missing_entries(zf, STORYJSON_REQUIRED_FILES)
    if STORYJSON_REVIEW_TABLE in missing:
        return missing

    try:
        table = json.loads(zf.read(STORYJSON_REVIEW_TABLE).decode("utf-8"))
    except (KeyError, json.JSONDecodeError, UnicodeDecodeError) as exc:
        return [*missing, f"{STORYJSON_REVIEW_TABLE} is unreadable: {exc}"]

    names = set(zf.namelist())
    story_paths = sorted({
        _story_path(story_key)
        for entry in table.values()
        for data in (entry.get("infoUnlockDatas") or [])
        if (story_key := data.get("storyTxt"))
    })
    missing.extend(path for path in story_paths if path not in names)
    return missing


GAMEDATA_EXCEL = ReleaseDatasetSpec(
    dataset_id="gamedata.excel",
    owner="3aKHP",
    repo="ArknightsGameData",
    asset_name="zh_CN-excel.zip",
    required_files=GAMEDATA_FILES,
)

STORY_ZH_CN = ReleaseDatasetSpec(
    dataset_id="story.zh_CN",
    owner="3aKHP",
    repo="ArknightsStoryJson",
    asset_name="zh_CN.zip",
    required_files=STORYJSON_REQUIRED_FILES,
)

GAMEDATA_LEVELS = ReleaseDatasetSpec(
    dataset_id="gamedata.levels",
    owner="3aKHP",
    repo="ArknightsGameData",
    asset_name="zh_CN-levels.zip",
    required_files=LEVELS_REQUIRED_FILES,
)
